torch/_fullgraph.py

```diff
@@ -0,0 +1,52 @@
+import dataclasses
+import logging
+import os
+from typing import Dict, List, Optional
+
+import torch
+import torch._inductor.package
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass
+class _Compilation:
+    paths: List[List[str]]
+
+
+class _Package:
+    def __init__(
+        self,
+        *,
+        path: Optional[str] = None,
+    ):
+        self.path = path or os.path.expandvars("/tmp/torchinductor_$USER/model.pt2")
+        if os.path.exists(self.path) and os.path.isdir(self.path):
+            raise RuntimeError(
+                f"File {self.path} already exists as a directory. Please specify a file path."
+            )
+        self._compilations: Dict[str, _Compilation] = {}
+
+    def add_aoti(self, name: str, path: List[str]) -> None:
+        if name not in self._compilations:
+            self._compilations[name] = _Compilation([])
+        else:
+            raise RuntimeError("Recompilation NYI")
+        self._compilations[name].paths.append(path)
+
+    def num_models(self) -> int:
+        return len(self._compilations)
+
+    def num_compilations(self, name: str) -> int:
+        return len(self._compilations[name].paths)
```

Comment:
Do we need to save multiple compilations in a single file?  This seems kind of weird if we require fullgraph.  The APIs would be simpler (no need for a name) if we had one package per compile.

Reply:
Two cases I can think of:
 1. We need a single package to share the weights from different compilations. E.g. prefill and decoding are compiled differently but they share the same set of weights.
 2. For each python callable, there could be multiple recompilations and we also need a way to map the user provided path to different paths for recompilations if we're want to save one file per compilation.

Reply:
Is case 2 supported in this PR?  For case 2 we could auto-generate the name based on the specializations done (we need to have those anyway).

Maybe we could support packages with some sort of container object that could be passed to save:
```
def save_compile_artifacts(self, path_or_package: Union[str, MultiFilePackage]):
   ...
```